### PR TITLE
RSDK-5606: Use viam logger in testing.

### PR DESCRIPTION
**Pull request by dgottlieb**

### CODE REVIEW

File: logging/impl.go [hunk 1]
```diff
@@ -35,34 +35,47 @@ type (
...
-func (impl *impl) AddAppender(appender Appender) {
-	impl.appenders = append(impl.appenders, appender)
+func (imp *impl) AddAppender(appender Appender) {
+	imp.appenders = append(imp.appenders, appender)
 }
 
-func (impl *impl) Desugar() *zap.Logger {
-	return nil
+func (imp *impl) Desugar() *zap.Logger {
+	return imp.AsZap().Desugar()
```
Comment (dgottlieb) [inline logging/impl.go:44]: Providing a real implementation

File: logging/impl.go [hunk 2]
```diff
@@ -35,34 +35,47 @@ type (
...
 }
 
-func (impl *impl) Level() zapcore.Level {
-	return zapcore.InfoLevel
+func (imp *impl) Level() zapcore.Level {
+	return imp.level.AsZap()
```
Comment (dgottlieb) [inline logging/impl.go:48]: Providing a real implementation

File: logging/impl.go [hunk 3]
```diff
@@ -35,34 +35,47 @@ type (
...
 }
 
-func (impl *impl) Named(name string) *zap.SugaredLogger {
-	return nil
+func (imp *impl) Sublogger(subname string) Logger {
```
Comment (dgottlieb) [inline logging/impl.go:51]: `Sublogger` is just a different name for `Named`, but it returns a `Logger` as opposed to a `zap.SugaredLogger`.

After removing `ZapCompatibleLogger`, we're free to go back to `Named` if we prefer that. But for this (and potentially some other things coming up trying to replace zap) we may need to "duplicate" functions with just their return value substituted.

File: logging/impl.go [hunk 4]
```diff
@@ -35,34 +35,47 @@ type (
...
+	return &impl{
+		name:      newName,
+		level:     imp.level,
+		appenders: imp.appenders,
+	}
 }
 
-func (impl *impl) Sync() error {
+func (imp *impl) Named(name string) *zap.SugaredLogger {
+	return imp.AsZap().Named(name)
```
Comment (dgottlieb) [inline logging/impl.go:65]: Providing a real implementation

File: logging/impl.go [hunk 5]
```diff
@@ -71,24 +84,42 @@ func (impl *impl) Sync() error {
 	return multierr.Combine(errs...)
 }
 
-func (impl *impl) With(args ...interface{}) *zap.SugaredLogger {
-	return nil
+func (imp *impl) With(args ...interface{}) *zap.SugaredLogger {
+	return imp.AsZap().With(args...)
```
Comment (dgottlieb) [inline logging/impl.go:80]: Providing a real implementation

File: logging/impl.go [hunk 6]
```diff
@@ -71,24 +84,42 @@ func (impl *impl) Sync() error {
...
 }
 
-func (impl *impl) WithOptions(opts ...zap.Option) *zap.SugaredLogger {
-	return nil
+func (imp *impl) WithOptions(opts ...zap.Option) *zap.SugaredLogger {
+	return imp.AsZap().WithOptions(opts...)
```
Comment (dgottlieb) [inline logging/impl.go:84]: Providing a real implementation

File: logging/impl.go [hunk 7]
```diff
@@ -71,24 +84,42 @@ func (impl *impl) Sync() error {
...
 }
 
-func (impl *impl) AsZap() *zap.SugaredLogger {
-	return NewLogger("").AsZap()
+func (imp *impl) AsZap() *zap.SugaredLogger {
+	// A hack to primarily copy zap's observer core objects from the viam logger back down to the
```
Comment (dgottlieb) [inline logging/impl.go:88]: A change worthy of understanding. Hopefully the comment here explains what's going on. But when reading the code-review top to bottom, this won't make much sense. One has to first look at and understand [our testing log logic](https://github.com/viamrobotics/rdk/pull/3185/files#diff-6bb8ce03ad8192446d441a3d85bf0fb8307e3d7829bc027785b35ce95868da70L79-R83) that uses an "observed logger".

File: logging/logging.go [hunk 8]
```diff
@@ -76,12 +75,12 @@ func NewTestLogger(tb testing.TB) Logger {
 
 // NewObservedTestLogger is like NewTestLogger but also saves logs to an in memory observer.
 func NewObservedTestLogger(tb testing.TB) (Logger, *observer.ObservedLogs) {
-	logger := zaptest.NewLogger(tb, zaptest.WrapOptions(zap.AddCaller()))
+	logger := NewViamLogger("")
+	logger.AddAppender(NewStdoutAppender())
 	observerCore, observedLogs := observer.New(zap.LevelEnablerFunc(zapcore.DebugLevel.Enabled))
```
Comment (dgottlieb) [inline logging/logging.go:80]: An `observerCore` [is a](https://github.com/uber-go/zap/blob/master/zapcore/core.go#L25-L45) `zapcore.Core`.

Notably it [satisfies](https://github.com/uber-go/zap/blob/master/zapcore/core.go#L40-L44) the [Appender interface](https://github.com/viamrobotics/rdk/blob/8356443673d4b7be9c34edf935981b4df19b82af/logging/impl.go#L16-L22) for writing log entries. And writing this comment I realized our `Appender` interface is totally in the wrong file.

The `observedLogs` is just a [slice of log entries](https://github.com/uber-go/zap/blob/87577d85d58b6d92d0158967df29303d04d30e36/zaptest/observer/observer.go#L36-L40) that tests can inspect.

File: module/modmanager/manager_test.go [hunk 9]
```diff
@@ -602,48 +602,3 @@ func TestDebugModule(t *testing.T) {
...
-	parentAddr, err := os.MkdirTemp("", "viam-test-*")
-	test.That(t, err, test.ShouldBeNil)
-	defer os.RemoveAll(parentAddr)
-	parentAddr += "/parent.sock"
-
-	mgr := NewManager(parentAddr, logger, modmanageroptions.Options{UntrustedEnv: false})
-
-	channel := make(chan struct{})
-	go func() {
-		err = mgr.Add(context.Background(), modCfg)
```
Comment (dgottlieb) [inline module/modmanager/manager_test.go:633]: Somehow the logging changes changed timing such that we hit the race condition writing to this err variable. This turned into a whole discussion about what the test is trying to accomplish. @benjirewis recommended removing the test.

File: robot/session_manager.go [hunk 10]
```diff
@@ -20,7 +20,7 @@ func NewSessionManager(robot Robot, heartbeatWindow time.Duration) *SessionManag
 	m := &SessionManager{
 		robot:             robot,
 		heartbeatWindow:   heartbeatWindow,
-		logger:            logging.FromZapCompatible(robot.Logger().Named("session_manager")),
+		logger:            robot.Logger().Sublogger("session_manager"),
```
Comment (dgottlieb) [inline robot/session_manager.go:23]: Now that we have a `Sublogger` method, we can keep a viam-logger -> viam-logger without losing information passing through a zap logger. A test failure caught this.

The "fix" to copy the observer cores from a viam-logger -> zap logger also would have fixed the test failure this change also fixes. But the goal is to only use zap when interfacing with other repositories. There may be other cases, but those will get cleaned up when changing viam in production to use the viam logger.

**Review by maximpertsov — COMMENTED** [review]
This change looks good! I just have one question regarding [copying zap's observer core](https://github.com/viamrobotics/rdk/pull/3185/files#r1380301063)

File: logging/impl.go (hunk 7 above)
Comment (dgottlieb) [reply logging/impl.go:88]: This is a great observation. The intention of this code is that it was only necessary for testing.

But this has made me realize that's not good enough. The way the code is right now, we lose the net logger (send logs to `app`) when calling this. I need to do something about that.

**Comment by dgottlieb** [comment]
Closing in favor of the bigger integration changes

**Pull request closed without merging**